tools/update.php

```diff
@@ -0,0 +1,747 @@
+        && (installAptPackages($requirements, $upgrade_mode = true))
+    ) {
+        die(
+            'Could not upgrade all required packages. Exiting.'
+            . PHP_EOL
+        );
+    }
+    return true;
+}
+
+/**
+ * Returns a list of SQL patch files to apply based on LORIS version
+ * differences.
+ *
+ * @param string $loris_root   The directory where LORIS is installed
+ * @param string $version_from Pre-update LORIS version
+ * @param string $version_to   Post-update LORIS version
+ *
+ * @return array Patch files to be applied
+ */
+function getPatchesFromVersion($loris_root, $version_from, $version_to) : array
+{
+    // Semantic versioning: 0 = major, 1 = minor, 2 = bugfix
+    define('MAJOR', 0);
+    define('MINOR', 1);
+    define('BUGFIX', 2);
+    // Convert string representation of version into arrays of ints
+    // TODO: check for dev branches
+    $from_versions = array_map('intval', explode('.', $version_from));
+    $to_versions   = array_map('intval', explode('.', $version_to));
+    // Calculate difference between old version and latest
+    $diff_major = $to_versions[MAJOR] - $from_versions[MAJOR];
+    $diff_minor = $to_versions[MINOR] - $from_versions[MINOR];
+    if ($diff_major < 0) {
+        echo '[!] Your version of LORIS is ahead of the latest release. If you'
+        . ' are not a developer working on the bleeding edge of LORIS, then '
+        . 'something has gone very wrong. No SQL patches will be displayed.'
+        . PHP_EOL;
+        return array();
+    }
+    echo "[**] Latest version $version_to is ahead of installed $version_from "
+        . "by $diff_major MAJOR release(s), $diff_minor MINOR release(s)."
+        . PHP_EOL;
+
+    // For every major version released between the version that is installed
+    // and the latest version, add the relevant patches if they begin with
+    // a number in that range. For example, if upgrading from 17 to 19, the
+    // array will contain all aptches beginning with "17", "18", or "19".
+    // i.e.:
+    //   [0] => /var/www/loris/SQL/Release_patches/17.0_To_18.0_upgrade_A.sql
+    //   [1] => /var/www/loris/SQL/Release_patches/17.0_To_18.0_upgrade_B.sql
+    //   [2] => /var/www/loris/SQL/Release_patches/18.0_To_19.0_upgrade.sql
+    //   [3] => /var/www/loris/SQL/Release_patches/19.0_To_19.1_upgrade.sql
+    // Range will run once when $start = $end. We want this in the case where
+    // there is no difference in major versions but there is a change in minor
+    // versions. The below loops will, in this case, match the minor release
+    // patches.
+    $start = $from_versions[MAJOR];
+    $end   = $start; // default case: there is a minor release difference
+    if ($diff_major > 0) {
+        $end = $to_versions[MAJOR];
+    }
+    $patch_dir           = $loris_root . 'SQL/Release_patches/';
+    $all_release_patches = glob($patch_dir . '*.sql');
+    $patches = [];
+    if ($diff_major > 0 || $diff_minor > 0) {
```

Comment:
@johnsaigle most of the time projects will need to apply project specif SQL in between 2 releases to prepare the database, this cannot be supported in this script

Reply:
>this cannot be supported in this script

What specifically can't be supported? I'm not sure I understand.

Reply:
Assume I'm 2 releases behind, for each release I have PRE and POST custom SQL patches designed for my project (thats usually the case for any project upgrading). how can I run them using your script as follows:
1. `PRE_19.0.sql` -> project patches
2. `18.0_To_19.0.sql` -> loris patches
3. `POST_19.0.sql` -> project patches
4. `PRE_20.0.sql` -> project patches
5. `19.0_to_20.0.sql` -> loris patches
6. `POST_20.0.sql` -> project patches

Reply:
I didn't know about this process for projects. 

If it's just one release you could do

1. Source `PRE` patch.
2. Run script.
3. Source `POST` patch.

I'm not sure how to handle the case where a project is more than 1 major release behind. If there is a standard location and naming convention for the files it would probably be possible to do these checks. Is that the case? Do you have any suggestions for how to incorporate this feature?

Reply:
As discussed in person, the script will be amended to not do SQL patching if the installed LORIS version is more than 1 major version behind the latest release, as it's difficult to automate and likely quite risky.

The script will display a message informing the user that the update process is complete in terms of src code, packages, etc. but that they will need to apply patches manually.